Step3 - 질문 삭제하기 리팩터링

mins99 opened the pull request:
안녕하세요 리뷰어님!
step3 구현하여 리뷰 요청 드립니다.

step2에서 피드백 주셨던 `answer`의 `deleted = true` 와 관련해서,
`getAnswers()` 로 리스트를 가져오던 로직을 삭제하면서 우선은 따로 구현은 하지 않았습니다.
그래도 `isOwner`로 체크하는 로직에서는 삭제된 답변인지에 대해 체크하도록 구현했습니다!

벌써 마지막 JPA 미션이네요~ 마지막까지 잘 부탁드립니다 👏

mins99 committed:
docs(README): 3단계 - 질문 삭제하기 리팩터링 요구 사항 정리

mins99 committed:
refactor(CannotDeleteException): RuntimeException를 상속하도록 변경

mins99 committed:
feat(Answers): List<Answer>를 가지는 일급 컬렉션 구현

- 답변 글의 모든 답변자가 같은 경우 삭제가 가능하다.
- 삭제결과 답변의 삭제 히스토리가 반환된다

mins99 committed:
test(AnswersTest): 답변 리스트 객체 테스트 추가

mins99 committed:
feat(Answer): 답변 객체 기능 추가

- set 메서드명 변경
- 답변 삭제 및 삭제 이력에 대한 정보를 DeleteHistory에 저장하여 반환하는 기능 추가

mins99 committed:
test(AnswerTest): 답변 객체 테스트 추가

mins99 committed:
chore(DeleteHistoryTest): 미사용 파일 삭제

mins99 committed:
feat(Question): 질문 삭제 기능 구현

- 질문 삭제를 통해 질문의 상태값을 삭제 상태로 변경한다
- 로그인 사용자와 질문한 사람이 같은 경우 삭제할 수 있다
- 질문자와 답변자가 다른 경우 답변을 삭제할 수 없다
- 질문의 삭제 및 삭제 이력에 대한 정보를 DeleteHistory에 저장하여 반환

mins99 committed:
test(QuestionTest): 질문 객체 테스트 추가 및 삭제

mins99 committed:
refactor(QnaService): deleteQuestion 리팩토링

- 도메인 영역에 핵심 비지니스를 구현

mins99 committed:
chore(QnaServiceTest): 미사용 라이브러리 삭제

mins99 committed:
fix(Answer): 이미 삭제된 답변인지 체크 여부 추가

mins99 commented on src/main/java/qna/domain/Answer.java:
```diff
@@ -92,8 +95,13 @@ public boolean isDeleted() {
 
-    public void setDeleted(boolean deleted) {
-        this.deleted = deleted;
+    public void delete() {
+        this.deleted = true;
+    }
```
삭제 로직에 있어서는 Answer 같은 경우 테스트를 고려해서 delete()를 만들고 호출하도록 했는데,
메서드 호출과 직접 값 변경에서 고민이 되었습니다 🥲 
리뷰어님은 이런 간단한 값 변경은 어떤 방식으로 하는 것을 선호하시는지 궁금합니다~

seokhongkim replied:
저도 `setDeleted` 보다는 `delete` 와 같이 객체에게 메세지를 보내는 것을 더 선호합니다.

추가적으로 작성자만 답변을 삭제할 수 있기 때문에, `delete` 요청을 보낼 때 작성자가 일치하는지 검증도 해야하지 않을까 싶습니다!

mins99 committed:
test(QuestionRepositoryTest): 삭제된 질문 조회 테스트 추가

seokhongkim commented on src/main/java/qna/domain/Answer.java:
```diff
@@ -61,6 +61,9 @@ public Answer(Long id, User writer, Question question, String contents) {
     }
 
     public boolean isOwner(User writer) {
+        if (isDeleted()) {
```
`isOwner` 는 작성자가 일치하는지만 검사할 것 같은데, 삭제인 경우에도 `true` 를 반환한다면 메소드명, 메소드에 기대하는 것 이상의 로직이 있는 것은 아닐까요?
다른 개발자가 작성자가 일치하는지 확인하려는 목적으로 이 메소드를 사용할 때 로직을 확인하지 않으면 문제가 생길 수 있을 것 같아요!

호출하는 곳에서 `deleted = false` 또는 `notDeleted` 인 `Answer` 만 필터링해서 작성자가 일치하는지 확인하는 건 어떨까요?

mins99 replied:
Answer에 loginUser를 넘기려면 `Question -> Answers -> Answer` 순서로 호출되어야하는데 그러지 않게 로직을 구현하려다보니 
Answer의 isOwner에 과도한 로직이 들어가게 된것 같습니다.
생각해보면 Answer도 개별로 삭제할 수 있으므로 Answer도 User 정보를 참조해야 하는게 맞는것 같습니다.

추가로 notDeleted인 Answer에 대해서만 삭제하는 로직도 추가해보겠습니다~

seokhongkim commented on src/main/java/qna/domain/Answer.java:
```diff
@@ -92,8 +95,13 @@ public boolean isDeleted() {
+        this.deleted = true;
+    }
+
+    public DeleteHistory deleteAnswer() {
```
`delete` 와 `deleteAnswer` 는 똑같은 역할을 할 것 같은데, 내부 로직이 다르네요.
메소드를 사용하는 클라이언트 입장(혹은 다른 개발자의 입장)에서는 어떤 메소드를 호출해야 할 지 구현을 알아야 할 것 같아요.
하나의 메소드로 통합해보는 건 어떨까요?

mins99 replied:
헉.. 다른 개발자 입장에서는 `delete()` 만 호출해도 되는걸로 오해를 할 수 있겠네요. 🥲
deleteAnswer를 통해서만 삭제하도록 수정하겠습니다!

seokhongkim commented on src/main/java/qna/service/QnaService.java:
```diff
@@ -39,28 +30,9 @@ public Question findQuestionById(Long id) {
 
     @Transactional
-    public void deleteQuestion(User loginUser, Long questionId) throws CannotDeleteException {
+    public void deleteQuestion(User loginUser, Long questionId) {
```
도메인에서 주요 비즈니스 로직을 구현해서 서비스가 굉장히 단순해졌네요! 👍

seokhongkim reviewed (changes requested):
민성님, JPA 3단계 미션도 잘 구현해주셨네요! 👍 
리뷰 코멘트 조금 남기고, 질문주신 부분에 대해 제 생각도 답변 작성했습니다.
리뷰 내용 반영 또는 코멘트 부탁드리며, 추가로 궁금하신 점 있으시면 깃헙 코멘트 또는 슬랙 DM으로 편하게 말씀해주세요. 😀

mins99 committed:
refactor(Answer): 답변 삭제 로직 수정

- Question에서 호출 하고 있었던 작성자 체크를 Answer가 하도록 변경
- 삭제된 답변인지 체크하는 로직도 유효성 검증 로직에서 하도록 변경

mins99 committed:
test(AnswerTest): 답변 삭제 로직 관련 테스트 추가

- deleteAnswer 파라미터 추가

mins99 committed:
refactor(Answers): 답변 리스트 대상 객체 수정

- 기존 모든 답변 리스트 대상 -> 삭제가 되지 않은 리스트로 변경
- 삭제 로직에서 loginUser를 넘기도록 수정

mins99 committed:
test(AnswersTest): 답변 리스트 테스트 추가

mins99 committed:
refactor(Question): deleteQuestion 개선

- Question에 대한 검증 로직만 수행하도록 변경
- Answer에 대한 검증 및 삭제는 answer가 함

mins99 committed:
test(QuestionTest): 질문 삭제 테스트 추가

mins99 commented:
안녕하세요 리뷰어님!
삭제 로직 수정하여 재 리뷰요청 드립니다~

1. Question - 질문에 대한 검증 후 삭제를 하며, 답변에 대한 검증 및 삭제는 답변이 합니다.
2. Answer - Question 과 동일하게 답변에 대한 검증 후 삭제를 합니다.
3. Answers - 질문 및 답변 삭제시 삭제가 되지 않은 답변에 대해서만 삭제할 수 있도록 리스트를 추출합니다.

삭제에 있어 찜찜한(?) 부분이 있었는데 리뷰어님의 피드백을 통해 정리가 되었습니다!
확인 부탁드립니다! : )

seokhongkim reviewed (approved):
민성님, JPA 3단계 미션 잘 진행해주셨습니다. 👍 
피드백 드린 내용도 잘 반영해주셨네요!
PR은 이만 머지하도록 하겠습니다.

JPA 미션 진행하느라 고생 많으셨습니다.
남은 교육 과정에서도 민성님께서 많이 학습하고 성장하시기를 응원하겠습니다. 💪
